### PR TITLE
Add get_shape_value_component() to FE_NedelecSZ

### PR DESCRIPTION
**Pull request by kinnewig**
This adds `get_shape_value_component()`, `get_shape_grad_component()` and `get_shape_grad_grad_component()` to `FE_NedelecSZ`.

Also, the code from `get_data()` was moved to the new function `evaluate()`, as we need that part to compute the above functions.

### CODE REVIEW

File: source/fe/fe_nedelec_sz.cc [hunk 1]
```diff
@@ -162,13 +162,25 @@ FE_NedelecSZ<dim, spacedim>::shape_value(const unsigned int /*i*/,
...
 {
-  // Not implemented yet:
-  DEAL_II_NOT_IMPLEMENTED();
-  return 0.;
+  AssertIndexRange(i, this->n_dofs_per_cell());
+  AssertIndexRange(component, dim);
+
+  std::unique_ptr<
+    typename dealii::FiniteElement<dim, spacedim>::InternalDataBase>
+    data_ptr = std::make_unique<InternalData>();
```
Comment (bangerth) [inline source/fe/fe_nedelec_sz.cc:174]: Oh, this is expensive. Would it be useful to mention that these functions are expensive in the documentation of this class?

File: source/fe/fe_nedelec_sz.cc [hunk 2]
```diff
@@ -211,46 +236,67 @@ FE_NedelecSZ<dim, spacedim>::shape_grad_grad(const unsigned int /*i*/,
...
+
+
+
+template <>
+void
+FE_NedelecSZ<2, 2>::evaluate(
+  const std::vector<Point<2>> p_list,
+  const UpdateFlags           update_flags,
+  std::unique_ptr<typename dealii::FiniteElement<2, 2>::InternalDataBase>
+    &data_ptr) const
```
Comment (bangerth) [inline source/fe/fe_nedelec_sz.cc:282]: The function used to be implemented for arbitrary dimensions, but here you restrict it to `dim==2`. Did we ever call the `dim!=2` case before?

File: source/fe/fe_nedelec_sz.cc [hunk 3]
```diff
@@ -264,1302 +310,1151 @@ FE_NedelecSZ<dim, spacedim>::get_data(
...
-                          data.shape_values[dof_index3_1][q][0] = polyy[j][0];
-                          data.shape_values[dof_index3_1][q][1] = 0.0;
+template <>
+void
+FE_NedelecSZ<3, 3>::evaluate(
+  const std::vector<Point<3>> p_list,
+  const UpdateFlags           update_flags,
+  std::unique_ptr<typename dealii::FiniteElement<3, 3>::InternalDataBase>
+    &data_ptr) const
+{
```
Comment (bangerth) [inline source/fe/fe_nedelec_sz.cc:722]: Oh, I see, you here split things off for the 3d case. Why is this necessary? I think you lose the `<2,3>` case this way.

Comment (kinnewig) [reply source/fe/fe_nedelec_sz.cc:722]: I just made this to get rid of `case(dim):`.

And in the file [`include/deal.II/fe/fe_nedelec_sz.h`](https://github.com/dealii/dealii/blob/31a1263477fc0d5b85bb7a2456ebb32744616593/include/deal.II/fe/fe_nedelec_sz.h#L86) there is the static assert: 

 ```
   static_assert(dim == spacedim,
                "FE_NedelecSZ is only implemented for dim==spacedim!");
```

So I haven´t toght about the `<2,3>`  case.
But with `<2,3>` in mind (eventhough it is not supported at the moment) it may be better to merge those two functions into one again.

What do you think?

File: source/fe/fe_nedelec_sz.cc (hunk 1 above)
Comment (kinnewig) [reply source/fe/fe_nedelec_sz.cc:174]: I've added a remark. And reworked the Doxygen description of those functions.

File: source/fe/fe_nedelec_sz.cc (hunk 3 above)
Comment (kinnewig) [reply source/fe/fe_nedelec_sz.cc]: I've added a commit that merges these evaluate functions into one.
I keep that as an extra commit for the moment so that it can be undone. But if we want to keep it that way, I can rebase that commit into the initial commit.

Comment (bangerth) [reply source/fe/fe_nedelec_sz.cc]: No, don't worry -- this is ok.

**Comment by bangerth** [comment]
12 of the NedelecSZ tests fail now:
```
[projectroot.tests.fe.fe_nedelec_sz_hanging_nodes_convergencedebug.serial.fe/fe_nedelec_sz_hanging_nodes_convergence.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_nedelec_sz_hanging_nodes_convergencedebug/serial/fe_fe_nedelec_sz_hanging_nodes_convergence_debug/)
[projectroot.tests.fe.fe_nedelec_sz_face_orientationdebug.serial.fe/fe_nedelec_sz_face_orientation.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_nedelec_sz_face_orientationdebug/serial/fe_fe_nedelec_sz_face_orientation_debug/)
[projectroot.tests.numerics.project_bv_curl_conf_05debug.serial.numerics/project_bv_curl_conf_05.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.numerics.project_bv_curl_conf_05debug/serial/numerics_project_bv_curl_conf_05_debug/)
[projectroot.tests.fe.fe_nedelec_sz_face_valuesdebug.serial.fe/fe_nedelec_sz_face_values.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_nedelec_sz_face_valuesdebug/serial/fe_fe_nedelec_sz_face_values_debug/)
[projectroot.tests.numerics.project_bv_curl_conf_complexdebug.serial.numerics/project_bv_curl_conf_complex.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.numerics.project_bv_curl_conf_complexdebug/serial/numerics_project_bv_curl_conf_complex_debug/)
[projectroot.tests.fe.fe_nedelec_sz_gradient_divergence_theoremdebug.serial.fe/fe_nedelec_sz_gradient_divergence_theorem.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_nedelec_sz_gradient_divergence_theoremdebug/serial/fe_fe_nedelec_sz_gradient_divergence_theorem_debug/)
[projectroot.tests.fe.fe_nedelec_sz_hessian_divergence_theoremdebug.serial.fe/fe_nedelec_sz_hessian_divergence_theorem.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_nedelec_sz_hessian_divergence_theoremdebug/serial/fe_fe_nedelec_sz_hessian_divergence_theorem_debug/)
[projectroot.tests.fe.fe_conformity_dim_3_fe_nedelec_szdebug.serial.fe/fe_conformity_dim_3_fe_nedelec_sz.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_conformity_dim_3_fe_nedelec_szdebug/serial/fe_fe_conformity_dim_3_fe_nedelec_sz_debug/)
[projectroot.tests.fe.fe_nedelec_sz_non_rect_facedebug.serial.fe/fe_nedelec_sz_non_rect_face.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_nedelec_sz_non_rect_facedebug/serial/fe_fe_nedelec_sz_non_rect_face_debug/)
[projectroot.tests.fe.fe_nedelec_sz_divergence_theorem_hanging_nodesdebug.serial.fe/fe_nedelec_sz_divergence_theorem_hanging_nodes.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_nedelec_sz_divergence_theorem_hanging_nodesdebug/serial/fe_fe_nedelec_sz_divergence_theorem_hanging_nodes_debug/)
[projectroot.tests.fe.get_fe_by_name_01debug.serial.fe/get_fe_by_name_01.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.get_fe_by_name_01debug/serial/fe_get_fe_by_name_01_debug/)
[projectroot.tests.fe.fe_nedelec_sz_hanging_nodes_on_inhomogeneous_boundarydebug.serial.fe/fe_nedelec_sz_hanging_nodes_on_inhomogeneous_boundary.debug](https://ci.tjhei.info/job/dealii-serial/job/PR-16887/5/testReport/junit/projectroot.tests.fe.fe_nedelec_sz_hanging_nodes_on_inhomogeneous_boundarydebug/serial/fe_fe_nedelec_sz_hanging_nodes_on_inhomogeneous_boundary_debug/)
```
Can you take a look?
